tiledb/api/c_api_test_support/testsupport_capi_array.h

```diff
@@ -0,0 +1,201 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines test support for the array section of the C API.
+ */
+
+#ifndef TILEDB_TESTSUPPORT_CAPI_ARRAY_H
+#define TILEDB_TESTSUPPORT_CAPI_ARRAY_H
+
+#include "testsupport_capi_context.h"
+#include "tiledb/api/c_api/array/array_api_external.h"
+#include "tiledb/api/c_api/array/array_api_internal.h"
+#include "tiledb/api/c_api/array_schema/array_schema_api_internal.h"
+#include "tiledb/api/c_api/attribute/attribute_api_internal.h"
+#include "tiledb/api/c_api/dimension/dimension_api_internal.h"
+#include "tiledb/api/c_api/domain/domain_api_internal.h"
+
+namespace tiledb::api::test_support {
+
+class ordinary_array_exception : public StatusException {
+ public:
+  explicit ordinary_array_exception(const std::string& message = "")
+      : StatusException("error creating test array; ", message) {
+  }
+};
+
+/**
+ * Base class for an ordinary array.
+ */
+class ordinary_array_without_schema {
+ private:
+  const char* array_uri = "unit_capi_array";
+
+ public:
+  ordinary_context context{};
+  tiledb_array_handle_t* array{nullptr};
+
+  ordinary_array_without_schema() {
+    auto rc = tiledb_array_alloc(context.context, array_uri, &array);
+    if (rc != TILEDB_OK) {
+      throw std::runtime_error("error allocating test array");
+    }
+    if (array == nullptr) {
+      throw std::logic_error(
+          "tiledb_array_alloc returned OK but without array");
+    }
+  }
+
+  ~ordinary_array_without_schema() {
+    tiledb_array_free(&array);
+  }
+
+  [[nodiscard]] tiledb_ctx_handle_t* ctx() const {
+    return context.context;
+  }
+
+  [[nodiscard]] const char* uri() const {
+    return array_uri;
+  }
+};
+
+/**
+ * An ordinary array with a fully-allocated array schema object.
+ */
+struct ordinary_array : public ordinary_array_without_schema {
+ private:
+  tiledb_dimension_t* dim;
+  tiledb_domain_t* domain;
+  tiledb_attribute_t* attr;
+  tiledb_array_schema_t* schema;
```

Comment:
>While this may have been practical at the time of writing the tests, there's now clearly a need for fully-defined schemas with attributes, dimensions, etc. It may be worthwhile file a followup to go through and make these test support classes more complete.

This.

That's exactly how these classes got written initial. I very purposefully did not move a bunch of them into general test support functions because they were immature.